lib/ukrust/rust_glue.c

```diff
@@ -0,0 +1,62 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Authors: Vlad-Andrei Badoiu <vlad_andrei.badoiu@upb.ro>
+ *
+ * Copyright (c) 2021, University Politehnica of Bucharest. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ */
+
+#include <bindings_helper.h>
+#include <uk/alloc.h>
+#include <uk/essentials.h>
+ #include <uk/arch/types.h>
+
+void *rust_sys_malloc(__sz size)
```

Comment:
I think it would be clean to return `NULL` if there is no allocator that we could ask for memory. If we have a LIBC that does not use ukalloc, we could forward the calls to libc. I suggest to implement the compile-time check if `LIBUKALLOC` is there, I leave it to you if you want to include the libC case.
```c
#if CONFIG_LIBUKALLOC
#include <uk/alloc.h>
#elif CONFIG_HAVE_LIBC
#include <stdlib.h>
#else /* !CONFIG_LIBUKALLOC && !CONFIG_HAVE_LIBC */
#include <uk/arch/types.h>
#endif

#if CONFIG_LIBUKALLOC
void __ukrust_sys_malloc(__sz size)
{
   return uk_malloc(uk_alloc_get_default(), size);
}

/* <..> */
#elif CONFIG_HAVE_LIBC
void __ukrust_sys_malloc(__sz size)
{
   return malloc(size);
}

/* <..> */
#else /* !CONFIG_LIBUKALLOC && !CONFIG_HAVE_LIBC */
void __ukrust_sys_malloc(__sz size)
{
   return __NULL;
}

/* <..> */
#endif
```
I also suggest to prefix all C symbols of your glue code with `__ukrust_` or `__ukrust_sys_`. Since this library does not have an `exportsyms.uk` file and you do not use `static`, all these symbols are on global scope. This is to avoid potential name clashes.